spring-integration-core/src/main/java/org/springframework/integration/gateway/MessagingGatewaySupport.java

```diff
@@ -333,6 +352,14 @@ private Object doSendAndReceive(Object object, boolean shouldConvert) {
 					error = ((ErrorMessage) reply).getPayload();
 				}
 			}
+			if (reply == null && this.errorOnTimeout) {
```

Comment:
That does not make sense. In terms of Messaging, of course.
If you have anything to return, that must be request/reply gateway, not one-way send-and-forget.

Therefore the `<int-http:inbound-gateway>` is still for you to specify your custom HTTP headers for response.
From other side you'd like to know what to place in the `payload` to make a proper response.
You can just simply return `HttpStatus` instance:
```
if (responseContent != null) {
	if (responseContent instanceof HttpStatus) {
		response.setStatusCode((HttpStatus) responseContent);
	}
	else {
		this.writeResponse(responseContent, response, request.getHeaders().getAccept());
	}
}
else {
	setStatusCodeIfNeeded(response);
}
```
We do nothing with the content if it is just `HttpStatus`. For your case it is exactly `OK(200, "OK")`.

BTW it isn't clear how your message goes to the `channel.fooService.fireEvent.response`, if " the service-activator has fired it will send null". No one Spring Integration component sends `null` to the channel. At least right now.